### ISSUE TITLE
Continue refactoring out common code from app layers

### ISSUE THREAD

**Issue by tmeasday** [storybooks/storybook#2636]
Following on from https://github.com/storybooks/storybook/pull/2241

Checklist:

- [x] Update RN
- [ ] Refactor RN more aggressively
- [ ] Refactor manager code
- [ ] Refactor server code
- [ ] Update tests/docs

**Comment by tmeasday** [comment]
@igor-dv you mentioned something about wanting to separate the angular part of `@storybook/core` -- can you go into a little more detail?

**Comment by alterx** [comment]
@tmeasday, there are several functions for generating dynamic components and metadata repeated through the angular app and the knobs addon. We should unify them so we don't repeat ourselves

**Comment by igor-dv** [comment]
👆  + after #2564 will be one more place - storyshots.

**Comment by tmeasday** [comment]
@alterx / @igor-dv do knobs and storyshots not ultimately depend on `@storybook/angular` though? In which case the common functionality could maybe just stay in that package?

**Comment by igor-dv** [comment]
I think none of the addons currently are dependant on any of the app packages. We can technically do it (like @Hypnosphi suggested - just import things from extraneous deps, without defining them in package.json), but in this case, it feels a bit dirty to me - because of the purpose of the "app" package.

**Comment by tmeasday** [comment]
I think we need to figure out the story on the way that addons import from the various app packages anyway, right? I mean it's pretty hacky in storyshots as it stands, it would be best to figure out a pattern as we take it further.

I guess what I am wondering is whether it helps for (say) `@storybook/addon-storyshots` to depend on `@storybook/angular-core` instead of depending on `@storybook/angular`. In both cases non-angular storyshots users are going to get angular dependencies that they don't want. So it will need to peer-depend or optional depend (or something else?) in any case.

**Comment by alterx** [comment]
Yeah, I think this was a point of discussion when we I initially started working on the knobs addon. At that point in time we wanted to keep their functionality separated, after all just because you use storybook it doesn't mean you'll want to use the knobs addon. I still feel the same way. We should split out the metadata/component/module annotation stuff from both the app and the addons; these things aren't directly related to the app or addons. They're just things we need in order to make storybook work with angular as it stands.

**Comment by igor-dv** [comment]
All this recent transition to the core can be considered a refactoring? Or you have something else in mind?

**Comment by danielduan** [comment]
Do you think splitting the RN app from the browser UI is something that could be added to the list of refactoring here?

Instead of the browser keeping track of the state, we should just run that code in RN. This would allow our app to run and ship independently.

**Comment by tmeasday** [comment]
@danielduan probably best to keep that to a separate issue. What I mean in this issue is quite a simple refactoring out of repeated code from the various app layers. It has been happening, slowly on both the preview and config (server) front.
